### PR TITLE
Better error message for calling set on a destroyed object.

### PR DESCRIPTION
**Pull request by jish**
cf: http://discuss.emberjs.com/t/could-calling-set-on-destroyed-object-error-be-more-descriptive/7730

### CODE REVIEW

File: packages/ember-metal/lib/property_set.js [hunk 1]
```diff
@@ -67,7 +67,7 @@ export function set(obj, keyName, value, tolerant) {
   }
 
   Ember.assert("You need to provide an object and key to `set`.", !!obj && keyName !== undefined);
-  Ember.assert('calling set on destroyed object', !obj.isDestroyed);
+  Ember.assert("calling set('" + keyName + "') on a destroyed object", !obj.isDestroyed);
```
Comment (rwjblue) [inline packages/ember-metal/lib/property_set.js]: can you also add 'obj' to the output. That will force its toString to be put into the assertion, which might help even more.

Comment (jish) [inline packages/ember-metal/lib/property_set.js]: Hrmm... it looks like I'm getting

screen shot 2015-06-09 at 5 44 43 pm

When I use:

```js
"') on a destroyed object (" + obj + ")"
```

And:

screen shot 2015-06-09 at 5 48 20 pm

When I use:

```js
"') on a destroyed object (" + obj.toString() + ")"
```

Comment (jamesarosen) [inline packages/ember-metal/lib/property_set.js]: Two things that do stringify well: `obj.constructor` and `Ember.guidFor(obj)`. I don't know if either is worth adding, but they might help.

**Comment by krisselden** [comment]
you can check if `typeof obj == 'object'` and if `typeof obj.toString == 'function'` is before call and if not you can do Object.prototype.toString.call(obj)  or String(obj)

**Comment by quaertym** [comment]
@krisselden :+1:

**Comment by jish** [comment]
So what's the story with `assert` here?

* Do we want to keep this as-is -- which is slightly more informative than it is currently?
* Do we want to put all of this `null` checking in around object?

If so, do we put all of the `null` checking inline?

```js
Ember.assert("calling set(" + (keyName && keyName.toString && typeof keyName.toString === 'function' ? "'" + keyName.toString() + "'" : 'undefined') + ") on a destroyed object (" + (obj && obj.toString && typeof obj.toString === 'function' ? obj.toString() : 'undefined') + ").", !obj.isDestroyed);
```

Or do we spread it out onto different lines? Will all that extra code be stripped properly for production builds?

**Comment by rwjblue** [comment]
> Do we want to put all of this null checking in around object?

Yes, I think so.

> If so, do we put all of the null checking inline?

Yes, so that it is stripped properly.

**Comment by rwjblue** [comment]
(also sorry for the delay in responding to your last question)

File: packages/ember-metal/lib/property_set.js (hunk 1 above)
Comment (stefanpenner) [inline packages/ember-metal/lib/property_set.js]: This should likely use a template string`${...}`

Comment (stefanpenner) [inline packages/ember-metal/lib/property_set.js]: > Two things that do stringify well: obj.constructor and Ember.guidFor(obj). I don't know if either is worth adding, but they might help.

obj.toString() should contain both of these values (for ember objects)

**Comment by jamesarosen** [comment]
@jish that's the longest line of code I've ever seen and liked :)

**Comment by jish** [comment]
Hi guys,

I've updated the pull request with your suggestions.

Unfortunately, a whole bunch of tests are failing for me locally today. They were all passing when I first made this pull request. I tried a little bit of guess and check with Travis CI... It looks like I'm not having much luck there either.

**Comment by jish** [comment]
Can someone help me with the failing test? There doesn't seem to be much debugging information in the Travis CI log:

https://travis-ci.org/emberjs/ember.js/jobs/67924624

    Running: ./node_modules/.bin/testem ci --port 7000
    ok 1 SL_Safari_Current - 

    ok 2 SL_Chrome_Current - 

    not ok 3 SL_IE_11 - 

    ok 4 SL_IE_10 - 

    ok 5 SL_IE_9 - 

    1..5
    # tests 5
    # pass  4
    # fail  1

**Comment by stefanpenner** [comment]
@jish looks like you will need to test your code manually in IE11

**Comment by trobrock** [comment]
Why was this closed? Did this get merged elsewhere @jish ?

**Comment by rwjblue** [comment]
I don't think this was merged, though it would have been a wonderful improvement.  

@jish  / @trobrock - Either of y'all care to pick this back up?

**Comment by oswaldoacauan** [comment]
:+1:

**Pull request closed without merging**